### PR TITLE
Add `torchmetrics`' own implementation of Rouge score metrics

### PR DESCRIPTION
**Pull request by stancld**
# Before submitting

- [x] Was this discussed/approved via a Github issue? (no need for typos and docs improvements)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), Pull Request section?
- [x] Did you make sure to update the docs?
- [x] Did you write any new necessary tests?

## What does this PR do?

This PR is a part of the endeavour of #433 to implement own metrics and minimize third-party dependencies. It adds the own implementation of Rouge score metrics.

In this PR, the `rouge-score` package (https://pypi.org/project/rouge-score/) is replaced with `torchmetrics`' own implementation. The behaviour of this new implementation should be equivalent tothe aforementioned package.

Furthermore, this PR strips the use of the `decimal_places` argument as this seems not to be used by any other metric. **TODO:** Add some depreciation warning (I will need a help here as I haven't found any own Warning class as it is the case in the `pytorch-lightning`.

**Future TODOs for another PR:** Replace `nltk` package (I will need to dig a bit deeper to find out if there's a way)

## PR review

Anyone in the community is free to review the PR once the tests have passed.
If we didn't discuss your PR in Github issues there's a high chance it will not be merged.

**Reviewers:** @Borda @SkafteNicki 

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: tests/text/test_rouge.py [hunk 1]
```diff
@@ -11,189 +11,179 @@
...
+    "rouge2_precision": torch.tensor(0.05),
+    "rouge2_recall": torch.tensor(0.05714285714285714),
+    "rouge2_fmeasure": torch.tensor(0.05333333333333333),
+    "rougeL_precision": torch.tensor(0.3111111111111111),
+    "rougeL_recall": torch.tensor(0.325),
+    "rougeL_fmeasure": torch.tensor(0.31764705882352945),
+    "rougeLsum_precision": torch.tensor(0.3111111111111111),
+    "rougeLsum_recall": torch.tensor(0.325),
+    "rougeLsum_fmeasure": torch.tensor(0.31764705882352945),
+}
```
Comment (justusschock) [inline tests/text/test_rouge.py:77]: Instead of hardcoding the results, could you maybe compare to the other rouge-score package?

File: torchmetrics/text/rouge.py [hunk 2]
```diff
@@ -111,11 +117,9 @@ def __init__(
                 raise ValueError(f"Got unknown rouge key {key}. Expected to be one of {ALLOWED_ROUGE_KEYS}")
 
         self.rouge_keys = rouge_keys
-        self.newline_sep = newline_sep
-        self.use_stemmer = use_stemmer
-        self.aggregator = BootstrapAggregator()
-        self.scorer = RougeScorer(rouge_keys, use_stemmer=self.use_stemmer)
-        self.decimal_places = decimal_places
+        self.rouge_keys_values = [ALLOWED_ROUGE_KEYS[key] for key in rouge_keys]
+        self.stemmer = nltk.stem.porter.PorterStemmer() if use_stemmer else None
```
Comment (justusschock) [inline torchmetrics/text/rouge.py:121]: is this the only part nltk is used? If so can we maybe make the import conditional of `use_stemmer`?

File: torchmetrics/functional/text/rouge.py [hunk 3]
```diff
@@ -12,114 +12,229 @@
...
+        stemmer:
+            Porter stemmer instance to strip word suffixes to improve matching.
 
     Example:
         >>> targets = "Is your name John".split()
         >>> preds = "My name is John".split()
-        >>> aggregator = BootstrapAggregator()
-        >>> scorer = RougeScorer(rouge_types=("rouge1", "rouge2", "rougeL", "rougeLsum"), use_stemmer=False)
-        >>> _rouge_score_update(preds, targets, scorer=scorer, aggregator=aggregator, newline_sep=False)
+        >>> from pprint import pprint
```
Comment (justusschock) [inline torchmetrics/functional/text/rouge.py:185]: I think this is not listed in the requirements?

File: torchmetrics/functional/text/rouge.py [hunk 4]
```diff
@@ -12,114 +12,229 @@
...
+    "rouge7": 7,
+    "rouge8": 8,
+    "rouge9": 9,
+    "rougeL": "L",
+    "rougeLsum": "Lsum",
+}
+
+
+@dataclass
+class _RougeScore:
```
Comment (justusschock) [inline torchmetrics/functional/text/rouge.py:43]: Why do you need a specific class for it? Doesn't a dict work here?

Comment (stancld) [reply torchmetrics/functional/text/rouge.py:43]: Yes, `dict` should be sufficient here. I will replace that!

File: torchmetrics/text/rouge.py (hunk 2 above)
Comment (stancld) [reply torchmetrics/text/rouge.py:121]: Actually, there are 2 places where the `nltk` package:
1. when `use_stemmer=True`
2. when `RougeLsum` is calculated.

Though, there are still plenty of cases where `nltk` is not required. Using `import nltk` conditionally is a good idea.

File: tests/text/test_rouge.py (hunk 1 above)
Comment (stancld) [reply tests/text/test_rouge.py:77]: Maybe, can I use the referred `rouge-score` package and add this one into the tests' requirements only?

Comment (justusschock) [reply tests/text/test_rouge.py:77]: exactly. That would be perfect, so that we also have tests if we decide to change the examples for tests.

File: torchmetrics/functional/text/rouge.py (hunk 3 above)
Comment (stancld) [reply torchmetrics/functional/text/rouge.py:177]: Actually, I haven't checked the requirements for the `pprint` as it has already been in use. But it is not listed anywhere in the requirements.

Comment (Borda) [reply torchmetrics/functional/text/rouge.py:174]: pprint is native (build-in) package

File: torchmetrics/functional/text/rouge.py [hunk 5]
```diff
@@ -12,114 +12,220 @@
...
     Example:
         >>> targets = "Is your name John".split()
         >>> preds = "My name is John".split()
-        >>> aggregator = BootstrapAggregator()
-        >>> scorer = RougeScorer(rouge_types=("rouge1", "rouge2", "rougeL", "rougeLsum"), use_stemmer=False)
-        >>> _rouge_score_update(preds, targets, scorer=scorer, aggregator=aggregator, newline_sep=False)
+        >>> from pprint import pprint
+        >>> pprint(
+        ...     _rouge_score_update(preds, targets, rouge_keys_values=[1, 2, 3, 'L'])
+        ... )  # doctest: +NORMALIZE_WHITESPACE +SKIP
```
Comment (Borda) [inline torchmetrics/functional/text/rouge.py:177]:
```suggestion
        >>> score = _rouge_score_update(preds, targets, rouge_keys_values=[1, 2, 3, 'L'])
        >>> pprint(score)  # doctest: +NORMALIZE_WHITESPACE +SKIP
```

File: torchmetrics/functional/text/rouge.py [hunk 6]
```diff
@@ -12,114 +12,220 @@
...
+            target_sum = _normalize_text(_add_newline_to_end_of_each_sentence(target_raw), stemmer)
+
+        for rouge_key in rouge_keys_values:
+            results[rouge_key].append(
+                _rouge_n_score(pred, target, rouge_key)
+                if isinstance(rouge_key, int)
+                else _rouge_l_score(
+                    pred if rouge_key != "Lsum" else pred_sum, target if rouge_key != "Lsum" else target_sum
+                )
+            )
```
Comment (Borda) [inline torchmetrics/functional/text/rouge.py:198]:
```suggestion
            if isinstance(rouge_key, int):
                score = _rouge_n_score(pred, target, rouge_key)
            else:
                score = _rouge_l_score(
                    pred if rouge_key != "Lsum" else pred_sum,
                    target if rouge_key != "Lsum" else target_sum,
                )
            results[rouge_key].append(score)
```

**Review by Borda — APPROVED** [review]

**Review by SeanNaren — APPROVED** [review]

File: torchmetrics/text/rouge.py [hunk 7]
```diff
@@ -17,28 +17,26 @@
...
-    from rouge_score.rouge_scorer import RougeScorer
-    from rouge_score.scoring import BootstrapAggregator
-else:
-    RougeScorer, BootstrapAggregator = object, object
+from torchmetrics.utilities.imports import _NLTK_AVAILABLE
 
 
 class ROUGEScore(Metric):
     """Calculate `ROUGE score <https://en.wikipedia.org/wiki/ROUGE_(metric)>`_, used for automatic summarization.
+    This implementation should imitathe the behaviour of the `rouge-score` package https://pypi.org/project/rouge-
```
Comment (SkafteNicki) [inline torchmetrics/text/rouge.py:25]:
```suggestion
    This implementation should imitate the behaviour of the `rouge-score` package https://pypi.org/project/rouge-
```

File: torchmetrics/text/rouge.py [hunk 8]
```diff
@@ -97,12 +95,13 @@ def __init__(
             process_group=process_group,
             dist_sync_fn=dist_sync_fn,
         )
+        # TODO: Add deprecated warning for newline_sep argument.
```
Comment (SkafteNicki) [inline torchmetrics/text/rouge.py:98]: Lets change the unused variables from
```
newline_sep: bool = False, 
```
to
```
newline_sep: Optional[bool] = None,
``` 
and then have a warning like this:
```suggestion
        if newline_sep is not None:
            warn.warning("Argument `newline_sep` is deprecated in v0.6 and will be removed in v0.7")
```

**Review by SkafteNicki — COMMENTED** [review]
Looks great :]
Remember to add note in changelog

**Review by justusschock — APPROVED** [review]

**Review by SkafteNicki — APPROVED** [review]
